simdev/module/github/gather_popular_repos_stage.py

```diff
@@ -0,0 +1,197 @@
+
+        # Cache the main fetching function
+        self._count_top_repos = PipelineCache.memory.cache(self._count_top_repos)
+
+    def run(self, pipeline: Pipeline):
+        """
+        First of all, we fetch all the stargazers for requested
+        repositories (merge them not to include duplicates).
+        Then we run over all the stargazers and fetch their
+        starred repositories (saving intermediate results if
+        possible). After that we accumulate all the results and
+        determine top-n repositories starred by initial
+        stargazers.
+        :param pipeline: pipline inside which the stage is run
+        """
+        self._context.popular_repositories = self._count_top_repos()
+
+    def _count_top_repos(self):
+        """
+        Firstly, we retrieve all the stargazers for the requested repositories,
+        merging them to avoid duplicates.
+        Next, we iterate through each stargazer and fetch their
+        starred repositories, saving intermediate results
+        whenever feasible.
+        Then, we gather all the results and identify the top-n
+        repositories that have been starred
+        by the original stargazers
+        :return: The list of top-n repositories, ranked by
+        the number of stars received from the initial stargazers
+        """
+        source_tqdm = tqdm(
+            self._context.source_repositories,
+            desc="Coursing through source repositories",
+        )
+        repo_star_counter = Counter()
+
+        # Acknowledged stargazers maintained to avoid duplicates between several
+        # source repositories
+        acked_stargazers: Set[str] = set()
+
+        for source_repo in source_tqdm:
+            stargazers_context = StargazersContext(repo=source_repo)
+            stargazers_context.stargazers = self._api_wrapper.fetch_stargazers(
+                repo=source_repo,
+                progress=source_tqdm,
+                page_limit=self._context.page_limit,
+            )
+
+            selection = stargazers_context.stargazers - acked_stargazers
+            with Pool(processes=self.stargazers_fetch_process_count) as pool:
+                starred_repos_progress = tqdm(
+                    pool.imap(self._process_stargazer, selection),
+                    total=len(selection),
+                    desc=f"Get starred repos for stargazers of {source_repo}",
+                )
+
+                for stargazer, starred_repos in starred_repos_progress:
+                    for repo in starred_repos:
+                        repo_star_counter.update([repo])
+
+            acked_stargazers.update(stargazers_context.stargazers)
+
+        # Calculating top-n repos by the number of stars given to them by stargazers
+        # of source repositories
+        return dict(
+            Counter(repo_star_counter).most_common(self._context.max_popular_repos)
+        )
+
+    def _process_stargazer(self, stargazer):
```

Comment:
I would recommend to cache `_process_stargazer` (if it's not yet cached - haven't gone through all PR yet) - it will save API calls to github API
OR `fetch_starred_repositories`

Reply:
`fetch_starred_repositories`'s cached - good